### PR TITLE
introduce device and PCI types

### PR DESCRIPTION
**Pull request by tmakatos**

### CODE REVIEW

File: lib/libvfio-user.c [hunk 1]
```diff
@@ -1288,10 +1295,16 @@ vfu_create_ctx(vfu_trans_t trans, const char *path, int flags, void *pvt)
         return NULL;
     }
 
+    if (dev_type != VFU_DEV_TYPE_PCI) {
+        errno = EINVAL;
```
Comment (swapnili) [inline lib/libvfio-user.c:1299]: Should this be ENOTSUP?

File: include/libvfio-user.h [hunk 2]
```diff
@@ -261,18 +268,32 @@ vfu_setup_log(vfu_ctx_t *vfu_ctx, vfu_log_fn_t *log, vfu_log_lvl_t log_lvl);
...
+ * @revision: PCI/PCI-X/PCIe revision
+ *
+ * @returns 0 on success, -1 on failure and sets errno.
  */
 int
 vfu_pci_setup_config_hdr(vfu_ctx_t *vfu_ctx, vfu_pci_hdr_id_t id,
                          vfu_pci_hdr_ss_t ss, vfu_pci_hdr_cc_t cc,
-                         bool extended);
+                         vfu_pci_type_t pci_type,
+                         int revision __attribute__((unused)));
```
Comment (swapnili) [inline include/libvfio-user.h:296]: Consider using UNUSED macro, good to be consistent across the code.

File: lib/libvfio-user.c [hunk 3]
```diff
@@ -1455,9 +1480,10 @@ vfu_setup_region(vfu_ctx_t *vfu_ctx, int region_idx, size_t size,
     switch(region_idx) {
     case VFU_PCI_DEV_BAR0_REGION_IDX ... VFU_PCI_DEV_VGA_REGION_IDX:
         // Validate the config region provided.
-        if (region_idx == VFU_PCI_DEV_CFG_REGION_IDX &&
-            !is_valid_pci_config_space_region(flags, size)) {
+        if (region_idx == VFU_PCI_DEV_CFG_REGION_IDX) {
+            if (flags == VFU_REG_FLAG_RW) {
```
Comment (jlevon) [inline lib/libvfio-user.c:1484]: what is this really testing for? Not using & for flags looks odd at least?

**Review by jlevon — CHANGES_REQUESTED** [review]

File: lib/libvfio-user.c (hunk 3 above)
Comment (tmakatos) [reply lib/libvfio-user.c:1484]: sorry that shoud be `!=`

Comment (tmakatos) [reply lib/libvfio-user.c:1484]: > what is this really testing for? Not using & for flags looks odd at least?

sorry that shoud be !=

**Comment by tmakatos** [comment]
@jlevon fixed, please take another look

File: lib/libvfio-user.c (hunk 1 above)
Comment (tmakatos) [reply lib/libvfio-user.c:1299]: No, `ENOTSUP` should be for thing we _could_ support but we don't. The user might set `dev_type` to something completely bogus.

**Review by jlevon — APPROVED** [review]

File: lib/libvfio-user.c (hunk 1 above)
Comment (swapnili) [reply lib/libvfio-user.c:1299]: But today we only support PCI and isn't the idea of adding device_type is to say we will support non-pci in future.

Comment (tmakatos) [reply lib/libvfio-user.c:1299]: When we add another device type we'll document it, so if the user supplies a wrong value then it's an invalid argument.

**Pull request merged**